File: apps/mobile/lib/core/utils/file_change.dart
import 'dart:convert';

import 'package:flutter/material.dart';

import '../theme/eco_theme.dart';

enum FileChangePreviewLineKind { add, remove, context }

class FileChangePreviewLine {
  const FileChangePreviewLine({required this.kind, required this.text});

  final FileChangePreviewLineKind kind;
  final String text;
}

class ThreadRunFileChangeMetadata {
  const ThreadRunFileChangeMetadata({
    required this.path,
    required this.additions,
    required this.deletions,
    required this.previewLines,
  });

  final String path;
  final int additions;
  final int deletions;
  final List<FileChangePreviewLine> previewLines;
}

/// Muted diff colors for activity file-change cards.
class FileChangeDiffPalette {
  const FileChangeDiffPalette({
    required this.addStat,
    required this.removeStat,
    required this.addGutter,
    required this.removeGutter,
    required this.addBackground,
    required this.removeBackground,
  });

  final Color addStat;
  final Color removeStat;
  final Color addGutter;
  final Color removeGutter;
  final Color addBackground;
  final Color removeBackground;

  factory FileChangeDiffPalette.of(EcoColors eco) {
    return FileChangeDiffPalette(
      addStat: eco.success.withValues(alpha: 0.58),
      removeStat: eco.danger.withValues(alpha: 0.58),
      addGutter: eco.success.withValues(alpha: 0.28),
      removeGutter: eco.danger.withValues(alpha: 0.28),
      addBackground: eco.success.withValues(alpha: 0.05),
      removeBackground: eco.danger.withValues(alpha: 0.05),
    );
  }

  Color gutterFor(FileChangePreviewLineKind kind) => switch (kind) {
        FileChangePreviewLineKind.add => addGutter,
        FileChangePreviewLineKind.remove => removeGutter,
        FileChangePreviewLineKind.context => Colors.transparent,
      };

  Color backgroundFor(FileChangePreviewLineKind kind) => switch (kind) {
        FileChangePreviewLineKind.add => addBackground,
        FileChangePreviewLineKind.remove => removeBackground,
        FileChangePreviewLineKind.context => Colors.transparent,
      };
}

class FileChangeCardDisplay {
  const FileChangeCardDisplay({
    required this.fileName,
    required this.path,
    required this.additions,
    required this.deletions,
    required this.previewLines,
  });

  final String fileName;
  final String path;
  final int additions;
  final int deletions;
  final List<FileChangePreviewLine> previewLines;
}

const _fileChangeTools = {'Edit', 'Write', 'MultiEdit', 'NotebookEdit'};

bool isFileChangeToolName(String toolName) => _fileChangeTools.contains(toolName);

ThreadRunFileChangeMetadata? resolveFileChangeFromToolInput(
  String toolName,
  Map<String, dynamic>? input,
) {
  if (!_fileChangeTools.contains(toolName) || input == null) {
    return null;
  }

  if (toolName == 'MultiEdit') {
    final edits = input['edits'];
    if (edits is List) {
      for (final edit in edits) {
        if (edit is! Map<String, dynamic>) continue;
        final resolved = _buildEditFileChange(
          _readString(edit['file_path']) ?? _readString(edit['path']),
          _readString(edit['old_string']),
          _readString(edit['new_string']),
        );
        if (resolved != null) return resolved;
      }
    }
    return null;
  }

  final filePath = _readString(input['file_path']) ??
      _readString(input['path']) ??
      _readString(input['notebook_path']);
  if (filePath == null) return null;

  if (toolName == 'Write') {
    return _buildWriteFileChange(filePath, _readString(input['content']) ?? '');
  }

  if (toolName == 'NotebookEdit') {
    return _buildWriteFileChange(
      filePath,
      _readString(input['new_source']) ?? _readString(input['source']) ?? '',
    );
  }

  return _buildEditFileChange(
    filePath,
    _readString(input['old_string']),
    _readString(input['new_string']),
  );
}

ThreadRunFileChangeMetadata? enrichFileChangeFromToolOutput(
  ThreadRunFileChangeMetadata? existing,
  Object? output,
) {
  final record = _parseToolOutputRecord(output);
  if (record == null) return existing;

  final filePath = _readString(record['filePath']) ??
      _readString(record['file_path']) ??
      _readString(record['notebook_path']) ??
      existing?.path;
  if (filePath == null) return existing;

  final gitDiff = record['gitDiff'];
  final structuredPatch = record['structuredPatch'];
  final fromPatch = structuredPatch is List
      ? _fileChangeFromStructuredPatch(structuredPatch)
      : null;

  final additions = gitDiff is Map
      ? (gitDiff['additions'] as num?)?.toInt() ?? fromPatch?.additions ?? existing?.additions ?? 0
      : fromPatch?.additions ?? existing?.additions ?? 0;
  final deletions = gitDiff is Map
      ? (gitDiff['deletions'] as num?)?.toInt() ?? fromPatch?.deletions ?? existing?.deletions ?? 0
      : fromPatch?.deletions ?? existing?.deletions ?? 0;
  final previewLines = fromPatch != null && fromPatch.previewLines.isNotEmpty
      ? fromPatch.previewLines
      : existing?.previewLines ?? const <FileChangePreviewLine>[];

  if (previewLines.isEmpty && additions == 0 && deletions == 0) {
    return existing;
  }

  return ThreadRunFileChangeMetadata(
    path: filePath,
    additions: additions,
    deletions: deletions,
    previewLines: previewLines,
  );
}

FileChangeCardDisplay? resolveFileChangeCardDisplay(
  ThreadRunFileChangeMetadata? metadata,
) {
  if (metadata == null || metadata.previewLines.isEmpty) return null;
  final normalizedPath = metadata.path.replaceAll('\\', '/');
  final fileName = normalizedPath.split('/').last;
  return FileChangeCardDisplay(
    fileName: fileName.isEmpty ? metadata.path : fileName,
    path: metadata.path,
    additions: metadata.additions,
    deletions: metadata.deletions,
    previewLines: metadata.previewLines,
  );
}

ThreadRunFileChangeMetadata? parseThreadRunFileChangeMetadata(Object? value) {
  if (value is! Map<String, dynamic>) return null;
  final path = _readString(value['path']);
  if (path == null) return null;
  final previewLines = _parsePreviewLines(value['previewLines']);
  if (previewLines.isEmpty) return null;
  return ThreadRunFileChangeMetadata(
    path: path,
    additions: (value['additions'] as num?)?.toInt() ?? 0,
    deletions: (value['deletions'] as num?)?.toInt() ?? 0,
    previewLines: previewLines,
  );
}

ThreadRunFileChangeMetadata? _buildEditFileChange(
  String? filePath,
  String? oldString,
  String? newString,
) {
  if (filePath == null) return null;
  final oldLines = _splitContentLines(oldString ?? '');
  final newLines = _splitContentLines(newString ?? '');
  final previewLines = <FileChangePreviewLine>[
    ...oldLines.map((text) => FileChangePreviewLine(kind: FileChangePreviewLineKind.remove, text: text)),
    ...newLines.map((text) => FileChangePreviewLine(kind: FileChangePreviewLineKind.add, text: text)),
  ];
  if (previewLines.isEmpty) return null;
  return ThreadRunFileChangeMetadata(
    path: filePath,
    additions: newLines.length,
    deletions: oldLines.length,
    previewLines: previewLines,
  );
}

ThreadRunFileChangeMetadata? _buildWriteFileChange(String filePath, String content) {
  final lines = _splitContentLines(content);
  if (lines.isEmpty) return null;
  return ThreadRunFileChangeMetadata(
    path: filePath,
    additions: lines.length,
    deletions: 0,
    previewLines: lines
        .map((text) => FileChangePreviewLine(kind: FileChangePreviewLineKind.add, text: text))
        .toList(),
  );
}

class _PatchSummary {
  const _PatchSummary({
    required this.additions,
    required this.deletions,
    required this.previewLines,
  });

  final int additions;
  final int deletions;
  final List<FileChangePreviewLine> previewLines;
}

_PatchSummary _fileChangeFromStructuredPatch(List<dynamic> patch) {
  final previewLines = <FileChangePreviewLine>[];
  var additions = 0;
  var deletions = 0;

  for (final hunk in patch) {
    if (hunk is! Map) continue;
    final lines = hunk['lines'];
    if (lines is! List) continue;
    for (final rawLine in lines) {
      if (rawLine is! String || rawLine.isEmpty) continue;
      final marker = rawLine[0];
      final text = rawLine.substring(1);
      switch (marker) {
        case '+':
          additions += 1;
          previewLines.add(FileChangePreviewLine(kind: FileChangePreviewLineKind.add, text: text));
        case '-':
          deletions += 1;
          previewLines.add(FileChangePreviewLine(kind: FileChangePreviewLineKind.remove, text: text));
        case ' ':
          previewLines.add(FileChangePreviewLine(kind: FileChangePreviewLineKind.context, text: text));
      }
    }
  }

  return _PatchSummary(
    additions: additions,
    deletions: deletions,
    previewLines: previewLines,
  );
}

Map<String, dynamic>? _parseToolOutputRecord(Object? output) {
  if (output is Map<String, dynamic>) return output;
  if (output is! String) return null;
  final trimmed = output.trim();
  if (!trimmed.startsWith('{') && !trimmed.startsWith('[')) return null;
  try {
    final parsed = _decodeJson(trimmed);
    return parsed is Map<String, dynamic> ? parsed : null;
  } catch (_) {
    return null;
  }
}

dynamic _decodeJson(String value) => jsonDecode(value);

List<FileChangePreviewLine> _parsePreviewLines(Object? value) {
  if (value is! List) return const [];
  final lines = <FileChangePreviewLine>[];
  for (final entry in value) {
    if (entry is! Map) continue;
    final text = entry['text'];
    final kind = entry['kind'];
    if (text is! String) continue;
    final parsedKind = switch (kind) {
      'add' => FileChangePreviewLineKind.add,
      'remove' => FileChangePreviewLineKind.remove,
      'context' => FileChangePreviewLineKind.context,
      _ => null,
    };
    if (parsedKind == null) continue;
    lines.add(FileChangePreviewLine(kind: parsedKind, text: text));
  }
  return lines;
}

List<String> _splitContentLines(String content) {
  if (content.isEmpty) return const [];
  return content.replaceAll('\r\n', '\n').split('\n');
}

String? _readString(Object? value) {
  if (value is! String || value.isEmpty) return null;
  return value;
}

File: apps/desktop/src/shared/file-change.ts
export type FileChangePreviewLineKind = "add" | "remove" | "context";

export interface FileChangePreviewLine {
  kind: FileChangePreviewLineKind;
  text: string;
}

export interface ThreadRunFileChangeMetadata {
  path: string;
  additions: number;
  deletions: number;
  previewLines: FileChangePreviewLine[];
}

export interface FileChangeCardDisplay {
  fileName: string;
  path: string;
  additions: number;
  deletions: number;
  previewLines: FileChangePreviewLine[];
}

const FILE_CHANGE_TOOLS = new Set(["edit", "write", "multiedit", "notebookedit"]);

export function isFileChangeToolName(toolName: string): boolean {
  return FILE_CHANGE_TOOLS.has(toolName.trim().toLowerCase());
}

export function resolveFileChangeFromToolInput(
  toolName: string,
  input: unknown,
): ThreadRunFileChangeMetadata | undefined {
  const normalized = toolName.trim().toLowerCase();
  if (!FILE_CHANGE_TOOLS.has(normalized) || !isRecord(input)) {
    return undefined;
  }

  if (normalized === "multiedit" && Array.isArray(input.edits)) {
    for (const edit of input.edits) {
      if (!isRecord(edit)) {
        continue;
      }
      const resolved = buildEditFileChange(
        readString(edit.file_path) ?? readString(edit.path),
        readString(edit.old_string),
        readString(edit.new_string),
      );
      if (resolved) {
        return resolved;
      }
    }
    return undefined;
  }

  const filePath = readString(input.file_path) ?? readString(input.path) ?? readString(input.notebook_path);
  if (!filePath) {
    return undefined;
  }

  if (normalized === "write") {
    return buildWriteFileChange(filePath, readString(input.content) ?? "");
  }

  if (normalized === "notebookedit") {
    return buildWriteFileChange(filePath, readString(input.new_source) ?? readString(input.source) ?? "");
  }

  const fromEdits = buildEditFileChangeFromPiEdits(filePath, input.edits);
  if (fromEdits) {
    return fromEdits;
  }

  return buildEditFileChange(
    filePath,
    readString(input.old_string) ?? readString(input.oldText) ?? readString(input.old_text),
    readString(input.new_string) ?? readString(input.newText) ?? readString(input.new_text),
  );
}

export function enrichFileChangeFromToolOutput(
  existing: ThreadRunFileChangeMetadata | undefined,
  output: unknown,
): ThreadRunFileChangeMetadata | undefined {
  const fromAcpDiff = fileChangeFromAcpDiffContent(output, existing?.path);
  if (fromAcpDiff && fromAcpDiff.previewLines.length > 0) {
    return fromAcpDiff;
  }

  const record = parseToolOutputRecord(output);
  if (!record) {
    return existing;
  }

  // ACP-style content may also be nested under a tool-result record.
  const nestedContent = record.content ?? record.contents;
  const fromNestedAcp = fileChangeFromAcpDiffContent(nestedContent, existing?.path);
  if (fromNestedAcp && fromNestedAcp.previewLines.length > 0) {
    return fromNestedAcp;
  }

  const filePath =
    readString(record.filePath) ??
    readString(record.file_path) ??
    readString(record.notebook_path) ??
    existing?.path;
  if (!filePath) {
    return existing;
  }

  const gitDiff = isRecord(record.gitDiff) ? record.gitDiff : undefined;
  const structuredPatch = Array.isArray(record.structuredPatch) ? record.structuredPatch : undefined;
  const fromPatch = structuredPatch ? fileChangeFromStructuredPatch(structuredPatch) : undefined;

  const additions =
    typeof gitDiff?.additions === "number"
      ? gitDiff.additions
      : (fromPatch?.additions ?? existing?.additions ?? 0);
  const deletions =
    typeof gitDiff?.deletions === "number"
      ? gitDiff.deletions
      : (fromPatch?.deletions ?? existing?.deletions ?? 0);
  const previewLines =
    fromPatch && fromPatch.previewLines.length > 0 ? fromPatch.previewLines : (existing?.previewLines ?? []);

  if (previewLines.length === 0 && additions === 0 && deletions === 0) {
    return existing;
  }

  return {
    path: filePath,
    additions,
    deletions,
    previewLines,
  };
}

/**
 * ACP tool_call content blocks: `{ type: "diff", path, oldText, newText }`.
 * Also accepts the nested `{ type: "content", content: { type: "diff", ... } }` form.
 */
function fileChangeFromAcpDiffContent(
  output: unknown,
  fallbackPath?: string,
): ThreadRunFileChangeMetadata | undefined {
  const diffs = collectAcpDiffBlocks(output);
  if (diffs.length === 0) {
    return undefined;
  }

  const filePath =
    diffs.find((entry) => entry.path)?.path ??
    (typeof fallbackPath === "string" && fallbackPath.trim() ? fallbackPath.trim() : undefined);
  if (!filePath) {
    return undefined;
  }

  const hasOld = diffs.some((entry) => entry.oldText !== undefined);
  const hasNew = diffs.some((entry) => entry.newText !== undefined);
  if (!hasOld && !hasNew) {
    return undefined;
  }

  return buildEditFileChange(
    filePath,
    hasOld ? diffs.map((entry) => entry.oldText ?? "").join("\n") : "",
    hasNew ? diffs.map((entry) => entry.newText ?? "").join("\n") : "",
  );
}

type AcpDiffBlock = {
  path?: string;
  oldText?: string | null;
  newText?: string | null;
};

function collectAcpDiffBlocks(content: unknown): AcpDiffBlock[] {
  if (Array.isArray(content)) {
    const blocks: AcpDiffBlock[] = [];
    for (const entry of content) {
      const block = readAcpDiffBlock(entry);
      if (block) {
        blocks.push(block);
      }
    }
    return blocks;
  }
  const single = readAcpDiffBlock(content);
  return single ? [single] : [];
}

function readAcpDiffBlock(value: unknown): AcpDiffBlock | undefined {
  if (!isRecord(value)) {
    return undefined;
  }
  if (value.type === "diff") {
    return normalizeAcpDiffBlock(value);
  }
  if (value.type === "content" && isRecord(value.content) && value.content.type === "diff") {
    return normalizeAcpDiffBlock(value.content);
  }
  return undefined;
}

function normalizeAcpDiffBlock(value: Record<string, unknown>): AcpDiffBlock | undefined {
  const path = readString(value.path) ?? readString(value.file_path) ?? readString(value.filePath);
  const hasOld = "oldText" in value || "old_text" in value || "old_string" in value;
  const hasNew = "newText" in value || "new_text" in value || "new_string" in value;
  if (!path && !hasOld && !hasNew) {
    return undefined;
  }
  return {
    ...(path ? { path } : {}),
    ...(hasOld ? { oldText: readNullableDiffText(value.oldText ?? value.old_text ?? value.old_string) } : {}),
    ...(hasNew ? { newText: readNullableDiffText(value.newText ?? value.new_text ?? value.new_string) } : {}),
  };
}

function readNullableDiffText(value: unknown): string | null {
  if (value === null) {
    return null;
  }
  return typeof value === "string" ? value : null;
}

export function resolveFileChangeCardDisplay(
  metadata: ThreadRunFileChangeMetadata | undefined,
): FileChangeCardDisplay | undefined {
  if (!metadata || metadata.previewLines.length === 0) {
    return undefined;
  }
  const normalizedPath = metadata.path.replace(/\\/g, "/");
  const fileName = normalizedPath.split("/").pop() || metadata.path;
  return {
    fileName,
    path: metadata.path,
    additions: metadata.additions,
    deletions: metadata.deletions,
    previewLines: metadata.previewLines,
  };
}

export function parseThreadRunFileChangeMetadata(value: unknown): ThreadRunFileChangeMetadata | undefined {
  if (!isRecord(value)) {
    return undefined;
  }
  const path = readString(value.path);
  if (!path) {
    return undefined;
  }
  const additions =
    typeof value.additions === "number" && Number.isFinite(value.additions) ? value.additions : 0;
  const deletions =
    typeof value.deletions === "number" && Number.isFinite(value.deletions) ? value.deletions : 0;
  const previewLines = parsePreviewLines(value.previewLines);
  return { path, additions, deletions, previewLines };
}

function parsePreviewLines(value: unknown): FileChangePreviewLine[] {
  if (!Array.isArray(value)) {
    return [];
  }
  const lines: FileChangePreviewLine[] = [];
  for (const entry of value) {
    if (!isRecord(entry) || typeof entry.text !== "string") {
      continue;
    }
    const kind = entry.kind;
    if (kind !== "add" && kind !== "remove" && kind !== "context") {
      continue;
    }
    lines.push({ kind, text: entry.text });
  }
  return lines;
}

function buildEditFileChangeFromPiEdits(
  filePath: string,
  edits: unknown,
): ThreadRunFileChangeMetadata | undefined {
  if (!Array.isArray(edits) || edits.length === 0) {
    return undefined;
  }
  const oldParts: string[] = [];
  const newParts: string[] = [];
  for (const edit of edits) {
    if (!isRecord(edit)) {
      continue;
    }
    const oldText = readString(edit.oldText) ?? readString(edit.old_string) ?? readString(edit.old_text);
    const newText = readString(edit.newText) ?? readString(edit.new_string) ?? readString(edit.new_text);
    if (oldText !== undefined) {
      oldParts.push(oldText);
    }
    if (newText !== undefined) {
      newParts.push(newText);
    }
  }
  if (oldParts.length === 0 && newParts.length === 0) {
    return undefined;
  }
  return buildEditFileChange(filePath, oldParts.join("\n"), newParts.join("\n"));
}

function buildEditFileChange(
  filePath: string | undefined,
  oldString: string | undefined,
  newString: string | undefined,
): ThreadRunFileChangeMetadata | undefined {
  if (!filePath) {
    return undefined;
  }
  const oldLines = splitContentLines(oldString ?? "");
  const newLines = splitContentLines(newString ?? "");
  const previewLines: FileChangePreviewLine[] = [
    ...oldLines.map((text) => ({ kind: "remove" as const, text })),
    ...newLines.map((text) => ({ kind: "add" as const, text })),
  ];
  if (previewLines.length === 0) {
    return undefined;
  }
  return {
    path: filePath,
    additions: newLines.length,
    deletions: oldLines.length,
    previewLines,
  };
}

function buildWriteFileChange(filePath: string, content: string): ThreadRunFileChangeMetadata | undefined {
  const lines = splitContentLines(content);
  if (lines.length === 0) {
    return undefined;
  }
  return {
    path: filePath,
    additions: lines.length,
    deletions: 0,
    previewLines: lines.map((text) => ({ kind: "add" as const, text })),
  };
}

function fileChangeFromStructuredPatch(
  patch: readonly unknown[],
): Pick<ThreadRunFileChangeMetadata, "additions" | "deletions" | "previewLines"> {
  const previewLines: FileChangePreviewLine[] = [];
  let additions = 0;
  let deletions = 0;

  for (const hunk of patch) {
    if (!isRecord(hunk) || !Array.isArray(hunk.lines)) {
      continue;
    }
    for (const rawLine of hunk.lines) {
      if (typeof rawLine !== "string" || rawLine.length === 0) {
        continue;
      }
      const marker = rawLine[0];
      const text = rawLine.slice(1);
      if (marker === "+") {
        additions += 1;
        previewLines.push({ kind: "add", text });
      } else if (marker === "-") {
        deletions += 1;
        previewLines.push({ kind: "remove", text });
      } else if (marker === " ") {
        previewLines.push({ kind: "context", text });
      }
    }
  }

  return { additions, deletions, previewLines };
}

function parseToolOutputRecord(output: unknown): Record<string, unknown> | undefined {
  if (isRecord(output)) {
    return output;
  }
  if (typeof output !== "string") {
    return undefined;
  }
  const trimmed = output.trim();
  if (!trimmed.startsWith("{") && !trimmed.startsWith("[")) {
    return undefined;
  }
  try {
    const parsed: unknown = JSON.parse(trimmed);
    return isRecord(parsed) ? parsed : undefined;
  } catch {
    return undefined;
  }
}

function splitContentLines(content: string): string[] {
  if (!content) {
    return [];
  }
  return content.replace(/\r\n/g, "\n").split("\n");
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function readString(value: unknown): string | undefined {
  return typeof value === "string" && value.length > 0 ? value : undefined;
}
